src/io/flutter/editor/FlutterColorProvider.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2021 The Chromium Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+package io.flutter.editor;
+
+import com.intellij.openapi.command.CommandProcessor;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.ElementColorProvider;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.impl.PsiFileFactoryImpl;
+import com.intellij.psi.impl.source.tree.AstBufferUtil;
+import com.jetbrains.lang.dart.DartLanguage;
+import com.jetbrains.lang.dart.DartTokenTypes;
+import com.jetbrains.lang.dart.psi.DartArgumentList;
+import com.jetbrains.lang.dart.psi.DartArguments;
+import com.jetbrains.lang.dart.psi.DartExpression;
+import com.jetbrains.lang.dart.psi.DartLiteralExpression;
+import io.flutter.FlutterBundle;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.awt.*;
+import java.util.Arrays;
+import java.util.List;
+
+import static io.flutter.dart.DartPsiUtil.getNewExprFromType;
+import static io.flutter.dart.DartPsiUtil.topmostReferenceExpression;
+
+public class FlutterColorProvider implements ElementColorProvider {
+
+  @Nullable
+  @Override
+  public Color getColorFrom(@NotNull PsiElement element) {
+    // This must return null for non-leaf nodes and any language other than Dart.
+    if (element.getNode().getElementType() != DartTokenTypes.IDENTIFIER) return null;
+
+    final String name = element.getText();
+    if (!(name.equals("Colors") || name.equals("CupertinoColors") || name.equals("Color"))) return null;
+
+    final PsiElement refExpr = topmostReferenceExpression(element);
+    if (refExpr == null) return null;
+    PsiElement parent = refExpr.getParent();
+    if (parent == null) return null;
+
+    if (parent.getNode().getElementType() == DartTokenTypes.ARRAY_ACCESS_EXPRESSION) {
+      // Colors.blue[200]
+      final String code = AstBufferUtil.getTextSkippingWhitespaceComments(parent.getNode());
+      return parseColorText(code.substring(code.indexOf(name) + name.length() + 1), name);
+    }
+    else if (parent.getNode().getElementType() == DartTokenTypes.CALL_EXPRESSION) {
+      // foo(Color.fromRGBO(0, 255, 0, 0.5))
+      return parseColorElements(parent, refExpr);
+    }
+    else if (parent.getNode().getElementType() == DartTokenTypes.SIMPLE_TYPE) {
+      // const Color.fromARGB(100, 255, 0, 0)
+      // parent.getParent().getParent() is a new expr
+      parent = getNewExprFromType(parent);
+      if (parent == null) return null;
+      return parseColorElements(parent, refExpr);
+    }
+    else {
+      // name.equals(refExpr.getFirstChild().getText()) -> Colors.blue
```

Comment:
I'm not sure what this comment and the next one express - is this condition more complex than when the color is presented in the form `Colors.blue`?

Reply:
These are "notes to self" (or next maintainer) about the structure of the PSI tree. JetBrains' Dart parser produces an AST in a style I remember reading about in my compilers class, but never actually used in any compiler I worked on. I work in the PSI tree every couple years, it seems, and it takes me some time to get back up to speed. These notes should help for next time.

In this case `name` is "Colors" and the first child of the reference expression is also "Colors" so we know we're looking at `Colors.blue` or some other named color. The next case is the structure for a shade reference, like `Colors.blue.shade200`.